File: nodeInsert.js
// Helper functions for insertion operations

function isLeftHull(sLower, sUpper) {
  if (sLower.type !== "segment" || sUpper.type !== "segment")
    throw "Invalid hull compare";
  var a = sUpper.a;
  var o = sLower.a;
  var b = sLower.b;
  var v0 = vec3(a[0] - o[0], a[1] - o[1], 0);
  var v1 = vec3(b[0] - o[0], b[1] - o[1], 0);
  return cross(v0, v1)[2] < 0;
}

function isClosing(child, p) {
  if (p.type !== "vec") throw "Invalid close test";

  var r = child.nextArc().site;
  var c = child.site;
  var l = child.prevArc().site;


  /* cases:
  1. l and c are segments and end at p
  2. c and r are segments and end at p
  2. l and r are segments and end at p
  */

  if (c.type === "segment" && r.type === "segment") {
    if (fastFloorEqual(r.b, p) && fastFloorEqual(c.b, p)) return {closeRight: true};
  }

  if (l.type === "segment" && c.type === "segment") {
    if (fastFloorEqual(l.b, p) && fastFloorEqual(c.b, p)) return {closeLeft: true};
  }

  // maybe is needed in the future
  // if (l.type === "segment" && r.type === "segment") {
  //   if (fastFloorEqual(l.b, p) && fastFloorEqual(r.b, p)) return true;
  // }
}

// function to create an edge node
function createNewEdge(left, right, vertex, dcel) {
  if (left.closeEvent) {
    left.closeEvent.live = false;
  }
  if (right.closeEvent) {
    right.closeEvent.live = false;
  }

  return new EdgeNode(left, right, vertex, dcel);
}

//------------------------------------------------------------
// Utility function right joint split
//  is an arc node. node is also an arc node.
//   \  |
//    \ |
//     \| node
//      *
//------------------------------------------------------------
function closePointSplit(left, right, dcel) {
  if (left.isV && right.isParabola) {
    return new EdgeNode(left, right, right.site, dcel);
  } else if (left.isParabola && right.isV) {
    return new EdgeNode(left, right, left.site, dcel);
  } else {
    throw 'invalid close joint split';
  }
}

//------------------------------------------------------------
// Utility function
// toSplit is an arc node. node is also an arc node.
//
//  |        | toSplit
//   \      /
//    |    ||
//     \__/ |
//          |
//          | node
//          *
//------------------------------------------------------------
function splitArcNode(toSplit, node, dcel, optNodesToClose) {
  if (toSplit.closeEvent) {
    toSplit.closeEvent.live = false;
  }
  var vertex;
  if (node.isV) {
    vertex = node.site.a;
  } else {
    var x = node.site[0];
    var y;
    if (toSplit.isParabola) {
      // avoid p==0 WATCH VALUE
      var directrix = toSplit.site[1] === node.site[1] ? node.site[1] - 1e-10: node.site[1];
      y = createParabola(toSplit.site, directrix, toSplit.id).f(x);
    } else {
      y = new V(toSplit.site, node.site[1], toSplit.id).f(x);
    }
    vertex = vec3(x, y, 0);
  }
  var left = toSplit;
  var right = new ArcNode(toSplit.site);
  if (optNodesToClose) {
    optNodesToClose.push(left);
    optNodesToClose.push(right);
  }
  return new EdgeNode(
      left, new EdgeNode(node, right, vertex, dcel), vertex, dcel);
}

function insertEdge(toSplit, edge, vertex, dcel, optNodesToClose) {
  if (toSplit.closeEvent) {
    toSplit.closeEvent.live = false;
  }

  var left = toSplit;
  var right = new ArcNode(toSplit.site);

  if (optNodesToClose) {
    optNodesToClose.push(left);
    optNodesToClose.push(right);
  }
  return new EdgeNode(
      left, new EdgeNode(edge, right, vertex, dcel), vertex, dcel);
}

// Child is guaranteed to be the parabola arc
function VRegularInsert(arcNode, childArcNode, dcel, parentV) {
  var left = isLeftHull(childArcNode.site, parentV.site);
  if (left) {
    // // Set edge information since we are using a left joint split
    var nextEdge = arcNode.nextEdge();
    if (nextEdge) nextEdge.dcelEdge.generalEdge = false;
    return createNewEdge(arcNode, childArcNode, childArcNode.site.a, dcel);
  } else {
    // // Set edge information since we are using a right joint split
    var prevEdge = arcNode.prevEdge();
    if (prevEdge) prevEdge.dcelEdge.generalEdge = false;
    // is a arc created by the right hull joint
    return createNewEdge(childArcNode, arcNode, childArcNode.site.a, dcel);
  }
}

function ParaInsert(child, arcNode, dcel, nodesToClose) {
  var newChild;

  var closingData = isClosing(child, arcNode.site);

  if (closingData) {
    if (!child.isV) throw "Invalid node insertion";
    // var sRight = child.nextArc();
    var edgeToUpdate = child.prevEdge();
    if (closingData.closeRight) {
      edgeToUpdate = child.nextEdge();
    }
    if (edgeToUpdate) {
      edgeToUpdate.dcelEdge.dest.overridden = true;
      edgeToUpdate.dcelEdge.dest.point = arcNode.site;
    }
    nodesToClose.push(child);
    if (closingData.closeRight) {
      nodesToClose.push(child.nextArc());
      newChild = closePointSplit(child, arcNode, dcel);
    } else {
      nodesToClose.push(child.prevArc());
      newChild = closePointSplit(arcNode, child, dcel);
    }
  } else {
    // regular split
    newChild = splitArcNode(child, arcNode, dcel);
    nodesToClose.push(arcNode.nextArc());
    nodesToClose.push(arcNode.prevArc());
  }
  return newChild;
}

function generateSubTree(eventPacket, arcNode, dcel, optChild) {
  var tree;
  var nodesToClose = [];
  var removeNode = undefined;
  var removePoint = undefined;
  if (eventPacket.type === PACKET_TYPE.MULTI_CHILD_PARENT) {
    leftArcNode = new ArcNode(eventPacket.leftChild);
    rightArcNode = new ArcNode(eventPacket.rightChild);
    var newEdge = createNewEdge(
        leftArcNode, rightArcNode, arcNode.site, dcel);
    if (optChild) {
      tree = splitArcNode(optChild, arcNode, dcel, nodesToClose);
      var parent = arcNode.parent;
      var childEdge =
          insertEdge(arcNode, newEdge, arcNode.site, dcel, nodesToClose);
      parent.setChild(childEdge, LEFT_CHILD);
    }
    else {
      tree = insertEdge(arcNode, newEdge, arcNode.site, dcel);
    }
  } else if (eventPacket.type === PACKET_TYPE.PARENT) {
    if (optChild) {
      if (!optChild.isV) throw 'Invalid insert operation';
      var childArcNode = new ArcNode(eventPacket.child);
      tree = splitArcNode(optChild, arcNode, dcel, nodesToClose);
      var parent = arcNode.parent;
      var newEdge = VRegularInsert(arcNode, childArcNode, dcel, optChild);
      parent.setChild(newEdge, LEFT_CHILD);
    } else {
      // case where site is the root
      var childArcNode = new ArcNode(eventPacket.child);
      tree = splitArcNode(arcNode, childArcNode, dcel, nodesToClose);
    }
  } else {
    if (optChild) {
      tree = ParaInsert(optChild, arcNode, dcel, nodesToClose);
    } else {
      tree = arcNode;
    }
  }

  return {root: tree, closingNodes: nodesToClose, optRemoveNode: removeNode, optRemovePoint: removePoint};
}

File: gvd2.0/nodeInsert.cc
#include "nodeInsert.hh"

#include "math.hh"

namespace
{
  bool isLeftHull(vec2 const& sLowerA, vec2 const& sLowerB, vec2 const& sUpperA)
  {
    auto v0 = vec2(sUpperA.x - sLowerA.x, sUpperA.y - sLowerA.y);
    auto v1 = vec2(sLowerB.x - sLowerA.x, sLowerB.y - sLowerA.y);
    return math::crossProduct(v0, v1) < 0;
  }

  // c++ form of isClosing()
  std::shared_ptr<bool> isClosingRight(std::shared_ptr<Node> child, vec2 const& p)
  {

    auto r = child->nextArc();
    // auto c = child.optSite;
    auto l = child->prevArc();

    if (!r || !l) return nullptr;

    // /* cases:
    // 1. l and c are segments and end at p
    // 2. c and r are segments and end at p
    // 2. l and r are segments and end at p
    // */

    if (child->aType == ArcType_e::ARC_V && r->aType == ArcType_e::ARC_V)
    {
      if (math::equiv2(r->b, p) && math::equiv2(child->b, p)) return std::make_shared<bool>(true);
    }

    if (l->aType == ArcType_e::ARC_V && child->aType == ArcType_e::ARC_V)
    {
      if (math::equiv2(l->b, p) && math::equiv2(child->b, p)) return std::make_shared<bool>(false);
    }

    // if (l.type === "segment" && c.type === "segment") {
    //   if (math::equiv2(l.b, p.point) && math::equiv2(c.b, p.point)) return std::make_shared<bool>(true);
    // }
    return nullptr;
  }

  std::shared_ptr<Node> createNewEdge(std::shared_ptr<Node> left, std::shared_ptr<Node> right, vec2 vertex,
                                      std::vector<CloseEvent>& rCQueue)
  {
    // left->live = false;
    // right->live = false;
    removeCloseEventFromQueue(left->id, rCQueue);
    removeCloseEventFromQueue(right->id, rCQueue);
    return math::createEdgeNode(left, right, vertex);
  }

  std::shared_ptr<Node> closePointSplit(std::shared_ptr<Node> left, std::shared_ptr<Node> right)
  {
    if (left->aType == ArcType_e::ARC_V && right->aType == ArcType_e::ARC_PARA)
    {
      return math::createEdgeNode(left, right, right->point);
    }
    else if (left->aType == ArcType_e::ARC_PARA && right->aType == ArcType_e::ARC_V)
    {
      return math::createEdgeNode(left, right, left->point);
    }

    throw std::runtime_error("Invalid close joint split");
    return nullptr;
  }

  std::shared_ptr<Node> splitArcNode(std::shared_ptr<Node> toSplit,
    std::shared_ptr<Node> node, std::vector<std::shared_ptr<Node>>& nodesToClose,
    std::vector<CloseEvent>& rCQueue)
  {
    // toSplit->live = false;
    removeCloseEventFromQueue(toSplit->id, rCQueue);
    vec2 vertex(0.0, 0.0);
    if (node->aType == ArcType_e::ARC_V)
    {
      vertex = node->a;
    }
    else
    {
      auto x = node->point.x;
      decimal_t y;
      if (toSplit->aType == ArcType_e::ARC_PARA)
      {
        auto d = toSplit->point.y == node->point.y ? node->point.y - 1e-10: node->point.y;
        auto h = toSplit->point.x;
        auto k = (d + toSplit->point.y) / 2;
        auto p = (toSplit->point.y - d) / 2;
        y = math::parabola_f(x, h, k, p);
      }
      else // else to split is a V
      {
        V obj(toSplit->a, toSplit->b, node->point.y, g_nodeId++);
        y = f_x(obj, x);
      }
      vertex = vec2(x, y);
    }
    auto eType = toSplit->aType == ArcType_e::ARC_PARA ? EventType_e::POINT : EventType_e::SEG;
    auto newEvent = eType == EventType_e::POINT ?
     Event(eType, toSplit->label, toSplit->point)
     : Event(eType, toSplit->label, vec2(0.0,0.0), toSplit->a, toSplit->b);
    auto right = math::createArcNode(newEvent);

    nodesToClose.push_back(toSplit);
    nodesToClose.push_back(right);
    auto rightEdge = math::createEdgeNode(node, right, vertex);
    return math::createEdgeNode(toSplit, rightEdge, vertex);
  }

  std::shared_ptr<Node> insertEdge(std::shared_ptr<Node> toSplit, std::shared_ptr<Node> edge,
        vec2 vertex, std::vector<std::shared_ptr<Node>>& nodesToClose,
        std::vector<CloseEvent>& rCQueue, bool addCloseNodes = true)
  {
    // toSplit->live = false;
    removeCloseEventFromQueue(toSplit->id, rCQueue);
    auto eType = toSplit->aType == ArcType_e::ARC_PARA ? EventType_e::POINT : EventType_e::SEG;
    auto newEvent = eType == EventType_e::POINT ?
     Event(eType, toSplit->label, toSplit->point)
     : Event(eType, toSplit->label, vec2(0.0,0.0), toSplit->a, toSplit->b);
    auto right = math::createArcNode(newEvent);
    if (addCloseNodes)
    {
      nodesToClose.push_back(toSplit);
      nodesToClose.push_back(right);
    }
    auto rightEdge = math::createEdgeNode(edge, right, vertex);
    return math::createEdgeNode(toSplit, rightEdge, vertex);
  }

  // Child is guaranteed to be the parabola arc
  std::shared_ptr<Node> VRegularInsert(std::shared_ptr<Node> arcNode,
              std::shared_ptr<Node> childArcNode, std::shared_ptr<Node> parentV,
              std::vector<CloseEvent>& rCQueue)
  {
    auto left = isLeftHull(childArcNode->a, childArcNode->b, parentV->a);
    if (left) {
      // // Set edge information since we are using a left joint split
      auto nextEdge = arcNode->nextEdge();
      // if (nextEdge) nextEdge.dcelEdge.generalEdge = false;
      return createNewEdge(arcNode, childArcNode, childArcNode->a, rCQueue);
    } else {
      // // Set edge information since we are using a right joint split
      auto prevEdge = arcNode->prevEdge();
      // if (prevEdge) prevEdge.dcelEdge.generalEdge = false;
      // is a arc created by the right hull joint
      return createNewEdge(childArcNode, arcNode, childArcNode->a, rCQueue);
    }
  }

  std::shared_ptr<Node> ParaInsert(std::shared_ptr<Node> child, std::shared_ptr<Node> arcNode,
                                  std::vector<std::shared_ptr<Node>>& nodesToClose,
                                  std::vector<CloseEvent>& rCQueue)
  {
    std::shared_ptr<Node> newChild = nullptr;
    // TODO performance - most nodes will not need this
    auto closingData = isClosingRight(child, arcNode->point);
    if (closingData)
    {
      // DEBUG ONLY
      if (child->aType != ArcType_e::ARC_V) throw std::runtime_error("Invalid node insertion");
      auto edgeToUpdate = child->prevEdge();
      if (*closingData)
      {
        edgeToUpdate = child->nextEdge();
      }
      if (edgeToUpdate)
      {
        edgeToUpdate->overridden = true;
        edgeToUpdate->edgeStart = arcNode->point; // General parabola points?
      }
      nodesToClose.push_back(child);
      if (*closingData)
      {
        nodesToClose.push_back(child->nextArc());
        newChild = closePointSplit(child, arcNode);
      }
      else
      {
        nodesToClose.push_back(child->prevArc());
        newChild = closePointSplit(arcNode, child);
      }
    }
    else
    {
      newChild = splitArcNode(child, arcNode, nodesToClose, rCQueue);
    }
    return newChild;
  }
}

SubTreeRslt generateSubTree(EventPacket const& e,
                                      std::shared_ptr<Node> arcNode,
                                      std::vector<CloseEvent>& rCQueue,
                                      std::shared_ptr<Node> optChild)
{
  auto tree = std::make_shared<Node>(ArcType_e::UNDEFINED, 0);
  std::vector<std::shared_ptr<Node>> nodesToClose;

  if (e.children.size() == 2)
  {
    auto leftArcNode = math::createArcNode(e.children[0]);
    auto rightArcNode = math::createArcNode(e.children[1]);
    auto newEdge = math::createEdgeNode(leftArcNode, rightArcNode, arcNode->point);
    if (optChild)
    {
      tree = splitArcNode(optChild, arcNode, nodesToClose, rCQueue);
      auto childEdge = insertEdge(arcNode, newEdge, arcNode->point, nodesToClose, rCQueue);
      math::setChild(tree->pRight, childEdge, Side_e::LEFT);
    }
    else
      tree = insertEdge(arcNode, newEdge, arcNode->point, nodesToClose, rCQueue, false);
  }
  else if (e.children.size() == 1)
  {
    if (optChild && optChild->aType == ArcType_e::ARC_V) {
      // if (!optChild.isV) throw 'Invalid insert operation';
      auto childArcNode = math::createArcNode(e.children[0]);
      tree = splitArcNode(optChild, arcNode, nodesToClose, rCQueue);
      auto parent = arcNode->pParent;
      auto newEdge = VRegularInsert(arcNode, childArcNode, optChild, rCQueue);
      math::setChild(parent, newEdge, Side_e::LEFT);
    } else if (optChild) {
      tree = splitArcNode(optChild, arcNode, nodesToClose, rCQueue);
      auto parent = arcNode->pParent;
      auto childArcNode = math::createArcNode(e.children[0]);
      auto newEdge = splitArcNode(arcNode, childArcNode, nodesToClose, rCQueue);
      math::setChild(parent, newEdge, Side_e::LEFT);
    } else {
      // case where site is the root
      auto childArcNode = math::createArcNode(e.children[0]);
      tree = splitArcNode(arcNode, childArcNode, nodesToClose, rCQueue);
    }
  }
  else
  {
    if (optChild)
      tree = ParaInsert(optChild, arcNode, nodesToClose, rCQueue);
    else
      tree = arcNode;
  }

 return {tree, nodesToClose};
}
